twelve-days: Implement canonical-data.json

alpakhomov opened the pull request:
------
Resolves #590 // m-a-age

alpakhomov commented:
I'm not sure about my implementation of canonical-data.json. The problem is that a lot of language tracks test every single line of the song - like 1, 2, 3 and so on till 12. I don't understand the reason behind it - it looks redundant to me. So, I think it's important to test the first and the last lines (checking the offset if some array-like data structure is used). Then, we check that all other parts of the song are there when we will run tests for intervals (for example, from first to third line). So, I don't really know what should be a priority - sticking to test cases that already are in many language tracks, or creating a suite of essential tests.
And maybe I'm wrong about this "test redundancy" in general.

Insti commented:
With only 12 items it doesn't seem unreasonable to test every verse individually.

alpakhomov commented:
Added test cases for every verse and changed their descriptions accordingly

Insti commented on exercises/twelve-days/canonical-data.json:
```diff
@@ -0,0 +1,96 @@
+  "version": "1.0.0",
+  "cases": [
+    {
+      "description": "returns verse number 1",
```
I'd like to see more descriptive descriptions. This does not tell me anything that cannot be automatically generated from the input.

How about:
"The first day's gift is a partridge in a pair tree"
"The second day's gift is two turtle doves"
etc.

Insti commented on exercises/twelve-days/canonical-data.json:
```diff
@@ -0,0 +1,96 @@
+    {
+      "description": "finds multiple verses, which are a fragment of the song",
+      "property": "verses",
+      "input": [1, 3],
```
There is no test for verses that doesn't start at 1
what should happen in that case?

alpakhomov commented:
Thanks @Insti, I've changed test descriptions and added test for interval from the middle of the song (4-6). Also changed interval test description accordingly.

I have also rebased commits into one.

alpakhomov commented:
Thank you @petertseng 
About this newlines - I just copied them from current implementations, it looks like most people included them in their tests. But I don't know if there is any logic behind it.

ErikSchierboom commented:
> I am wondering whether the trailing newlines on all expected outputs are necessary. Naively I might assume they aren't, but maybe there is something I don't know.

I agree. It makes sense for the verses to be separated by newlines, but for an individual verse it doesn't IMHO. Maybe this is the point where we should remove those extra newlines?

alpakhomov commented:
Okay, I a bit confused now, so, just to be clear - we are talking about last newline symbol at the end of each verse/block of verses, right?

Insti commented on exercises/twelve-days/canonical-data.json:
```diff
@@ -0,0 +1,102 @@
+      "description": "first day a partridge in a pear tree",
+      "property": "verse",
+      "input": 1,
+      "expected": "On the first day of Christmas my true love gave to me, a Partridge in a Pear Tree.\n"
```
I believe the suggestion is that this line should not have the `\n` at the end.

ErikSchierboom replied:
Correct

alpakhomov replied:
Removed trailing newline for single verse tests

Insti commented on exercises/twelve-days/canonical-data.json:
```diff
@@ -0,0 +1,102 @@
+      "description": "finds multiple verses, which are a fragment from the beggining of the song",
+      "property": "verses",
+      "input": [1, 3],
+      "expected": "On the first day of Christmas my true love gave to me, a Partridge in a Pear Tree.\n\nOn the second day of Christmas my true love gave to me, two Turtle Doves, and a Partridge in a Pear Tree.\n\nOn the third day of Christmas my true love gave to me, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree.\n"
```
But whenever there are multiple verses they should be separated by `\n\n`
But no trailing newline.

ErikSchierboom replied:
But it only makes sense in the context of multiple verses, so why add it to a single verse?

Insti replied:
> Maybe the trailing single newline should be there.

In this case I am wondering about how many newlines should be at the end of the completed `verses` string. 0, 1 or 2? Currently there is 1.

ErikSchierboom replied:
I would expect the following to hold:

1. A single verse would have no newline at the end
1. Verses (more than one) are separated by an empty line
1. Verses don't have trailing newlines

alpakhomov commented:
> A single verse would have no newline at the end
Verses (more than one) are separated by an empty line
Verses don't have trailing newlines (so that means 0)

Fixed

Insti commented:
Do you have an automated checker for this exercise @petertseng ?

petertseng commented on exercises/twelve-days/canonical-data.json:
```diff
@@ -0,0 +1,102 @@
+    {
+      "description": "finds multiple verses, which are a fragment from the beggining of the song",
+      "property": "verses",
+      "input": [1, 3],
```
hmm, sorry to swoop in with another  thing again, but I just remembered... what about assigning meaning to the two elements of this array? See how `beer song` does it at https://github.com/exercism/problem-specifications/blob/master/exercises/beer-song/canonical-data.json#L54-L55 . Any desire to do the same here?

alpakhomov replied:
Added two elements instead of one array; named them `beginning` and `end` respectively. Not sure if it will be a good idea to split all cases into two groups, like "beer-song" does (for single and multiple verse test cases).

Insti replied:
should `beginning` and `end` be subkeys of input?
```
"input": {"beginning": 1, "end": 3}
```

alpakhomov replied:
Should it? Wouldn't it be confused with some object or map data structure?

petertseng replied:
> Should `beginning` and `end` be subkeys of input?

That would be an interesting change to make for the JSON Files all around. Then, instead of having to vary the input keys per exercise, test generators would know that all inputs can be found in `"input"`. That could be a useful change for the schema! I think the track maintainer who wishes that to happen for all JSON files should make the issue requesting it.

rpottsoh commented on exercises/twelve-days/canonical-data.json:
```diff
@@ -0,0 +1,105 @@
+      "end": 3,
+      "expected": "On the first day of Christmas my true love gave to me, a Partridge in a Pear Tree.\n\nOn the second day of Christmas my true love gave to me, two Turtle Doves, and a Partridge in a Pear Tree.\n\nOn the third day of Christmas my true love gave to me, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree."
+    },
+    
```
Line 84 should be removed.

rpottsoh commented on exercises/twelve-days/canonical-data.json:
```diff
@@ -0,0 +1,105 @@
+    {
+      "description": "twelfth day twelve drummers drumming",
+      "property": "verse",
+      "input": 12,
```
Instead of `"input"` how about `"number"` since it is the verse number being asked for.

rpottsoh commented:
If you would like to make this more `beer-song` like, you could group the `Verse` cases under a description of "Verse" and all the other cases under a description of "Lyrics".

alpakhomov commented:
Thanks @rpottsoh 
I've splitted cases into two groups (`verse` and `lyrics`), renamed `input` to `number`, and removed unnecessary empty line.

petertseng commented:
> Do you have an automated checker for this exercise

OK, guess that wasn't so much to write.

```ruby
require 'json'
require_relative '../../verify'

DAYS = {
  twelfth: 'twelve Drummers Drumming',
  eleventh: 'eleven Pipers Piping',
  tenth: 'ten Lords-a-Leaping',
  ninth: 'nine Ladies Dancing',
  eighth: 'eight Maids-a-Milking',
  seventh: 'seven Swans-a-Swimming',
  sixth: 'six Geese-a-Laying',
  fifth: 'five Gold Rings',
  fourth: 'four Calling Birds',
  third: 'three French Hens',
  second: 'two Turtle Doves',
  first: 'a Partridge in a Pear Tree',
}.each_value(&:freeze).freeze

def list(things)
  things.size == 1 ? things[0] : (things[0...-1] << "and #{things[-1]}").join(', ')
end

def verse(n)
  ordinal = DAYS.keys[-n]
  gifts = DAYS.values.last(n)
  "On the #{ordinal} day of Christmas my true love gave to me, #{list(gifts)}."
end

def verses(a, b)
  (a..b).map { |n| verse(n) }.join("\n\n")
end

json = JSON.parse(File.read(File.join(__dir__, 'canonical-data.json')))

raise "There should be exactly two cases, not #{json['cases'].size}" if json['cases'].size != 2

verify(json['cases'][0]['cases'], property: 'verse') { |c|
  verse(c['number'])
}

subcases = by_property(json['cases'][1]['cases'], %w(verses sing))

verify(subcases['verses'], property: 'verses') { |c|
  verses(c['beginning'], c['end'])
}

verify(subcases['sing'], property: 'sing') { |c|
  verses(1, DAYS.size)
}
```

As written, it's fine.

Note that it may be a bit strange that both `verses` and `sing` are in the same group. As you can see though it can be dealt with.

Insti commented:
Should `verses` and `sing` have their own sections?

Insti commented:
Looking at the [description](https://github.com/exercism/problem-specifications/blob/master/exercises/twelve-days/description.md) and the policy that we should not be testing intermediate values it would seem that there should only be one test for the complete song.

alpakhomov commented:
I changed properties and descriptions of tests in the `lyrics` block - now there is only one test for the whole song, plus we can still have only two groups of test cases (`verse` and `lyrics`). I think it looks better now - but I can revert chages back if needed. 
About changes in exercise [description](https://github.com/exercism/problem-specifications/blob/master/exercises/twelve-days/description.md) - should it be done? If yes, do I need a separate pull request?

petertseng commented:
If we take this PR as is, this exercise is on par with:

* beer-song
  * whose https://github.com/exercism/problem-specifications/blob/master/exercises/beer-song/description.md only says to "produce the lyrics"
  * but whose https://github.com/exercism/problem-specifications/blob/master/exercises/beer-song/canonical-data.json checks individual verses
* house
  * whose https://github.com/exercism/problem-specifications/blob/master/exercises/house/description.md only says to "output the nursery rhyme"
  * but whose https://github.com/exercism/problem-specifications/blob/master/exercises/house/canonical-data.json checks individual verses
* food-chain
  * whose https://github.com/exercism/problem-specifications/blob/master/exercises/food-chain/description.md only says to "generate the lyrics"
  * but whose https://github.com/exercism/problem-specifications/blob/master/exercises/food-chain/canonical-data.json checks individual verses

Whatever treatment is given to this should also be given to all the others.

Insti commented:
Hi @alpakhomov, are you willing to make some more changes to this to match what we're planning on doing to food-chain in #973

`property`: was `verse` and `sing`, both become `recite`

`number`,`beginning`, `end`:  become `start verse` and `end verse`, each test case should have both. Where there was only one number they should `start verse` and `end verse` should be the same.

alpakhomov commented:
Changed property and parameter names, everything else is the same

ErikSchierboom commented:
Great work @alpakhomov ! What I was wondering, is if we should encoding the expected value as an array of strings instead of a single string, similar to what happens in the [food-chain canonical data](https://github.com/exercism/problem-specifications/blob/master/exercises/food-chain/canonical-data.json)? My personal preference would be to use an array. @Insti, what are your thoughts?

Insti commented:
> encoding the expected value as an array of strings

I have no strong opinion, an array of strings seems fine, so long as there is a comment about it for the test implementer: https://github.com/exercism/problem-specifications/blob/master/exercises/food-chain/canonical-data.json#L4-L7

ErikSchierboom commented on exercises/twelve-days/canonical-data.json:
```diff
@@ -0,0 +1,121 @@
+        {
+          "description": "first day a partridge in a pear tree",
+          "property": "recite",
+          "start verse": 1,
```
We have now come up with an updated format, which requires the properties to be lowerCamelCase (see #983). Could you update the property to `startVerse`? And do the same for the other "start verse" properties?

ErikSchierboom commented on exercises/twelve-days/canonical-data.json:
```diff
@@ -0,0 +1,121 @@
+          "description": "first day a partridge in a pear tree",
+          "property": "recite",
+          "start verse": 1,
+          "end verse": 1,
```
We have now come up with an updated format, which requires the properties to be lowerCamelCase (see #983). Could you update the property to `endVerse`? And do the same for the other "end verse" properties?

ErikSchierboom commented on exercises/twelve-days/canonical-data.json:
```diff
@@ -0,0 +1,121 @@
+          "property": "recite",
+          "start verse": 2,
+          "end verse": 2,
+          "expected": "On the second day of Christmas my true love gave to me, two Turtle Doves, and a Partridge in a Pear Tree."
```
Could you encode the expected value as an array of strings? See #983 and #973. And do the same for the other "expected" properties?

ErikSchierboom commented:
@alpakhomov Sorry about the churn in the changing requirements! It has taken us a little while to come up with uniform specifications, but the final result is in #983 and it should now be stable.

alpakhomov commented:
Renamed parameters with camelCase, changed all expected values to arrays (even single line ones), added comment about expected values same as here https://github.com/exercism/problem-specifications/blob/master/exercises/food-chain/canonical-data.json#L4-L7

ErikSchierboom commented:
I think this is ready to merge. @Insti, do you agree?

Insti commented:
Yes it looks good to me.

ErikSchierboom commented:
Merged! Thanks a lot @alpakhomov!

Insti commented:
Thanks for all your work on this @alpakhomov and your patience with the constantly changing requirements :heart:
